vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go

```diff
@@ -199,6 +204,65 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		go finalizingController.Run(5, context.StopCh)
 		return nil
 	})
+	s.GenericAPIServer.AddPostStartHookOrDie("crd-discovery-available", func(context genericapiserver.PostStartHookContext) error {
+		return wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
+			// only check if we have a valid list for a given resourceversion
+			if !s.Informers.Apiextensions().InternalVersion().CustomResourceDefinitions().Informer().HasSynced() {
+				return false, nil
+			}
+
+			_, resourceLists, err := crdClient.Discovery().ServerGroupsAndResources()
+			if err != nil {
+				return false, err
+			}
+			crds, err := s.Informers.Apiextensions().InternalVersion().CustomResourceDefinitions().Lister().List(labels.Everything())
+			if err != nil {
+				return false, err
+			}
+			missingCRDs := sets.String{}
+			for _, crd := range crds {
+				// only check for those that should be in discovery
+				if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) {
+					continue
+				}
+
+				for _, versionSpec := range crd.Spec.Versions {
+					foundCRDVersion := false
+					for _, resourceList := range resourceLists {
+						for _, resource := range resourceList.APIResources {
+							if resource.Group != crd.Spec.Group {
+								continue
+							}
+							if resource.Name != crd.Spec.Names.Plural {
+								continue
+							}
+							if resource.Version != versionSpec.Name {
+								continue
+							}
+
+							foundCRDVersion = true
+							break
+						}
+
+						if foundCRDVersion {
+							break
+						}
+					}
+
+					if !foundCRDVersion {
+						missingCRDs.Insert(crd.Spec.Names.Plural + "." + versionSpec.Name + "." + crd.Spec.Group)
+					}
+				}
+			}
+
+			if len(missingCRDs) > 0 {
+				glog.Warningf("waiting on CRD discovery for %v", missingCRDs.List())
```

Comment:
info

Reply:
I would want this to be clearly visible in the log (so we can confirm we indeed see missing CRD and without this hook we will return 404 to clients (and GC))